v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -0,0 +1,627 @@
+
+    client.set_credentials(username, password)
+
+    # Add topic filters for receive
+    # IoTHub Receives
+    c2d_msg_topic = mqtt_topic.get_c2d_topic_for_subscribe(client_config.device_id)
+    client.add_incoming_message_filter(c2d_msg_topic)
+    if client_config.module_id:
+        input_msg_topic = mqtt_topic.get_input_topic_for_subscribe(
+            client_config.device_id, client_config.module_id
+        )
+        client.add_incoming_message_filter(input_msg_topic)
+    method_request_topic = mqtt_topic.get_method_topic_for_subscribe()
+    client.add_incoming_message_filter(method_request_topic)
+    twin_patch_topic = mqtt_topic.get_twin_patch_topic_for_subscribe()
+    client.add_incoming_message_filter(twin_patch_topic)
+    # Operation Responses
+    twin_response_topic = mqtt_topic.get_twin_response_topic_for_subscribe()
+    client.add_incoming_message_filter(twin_response_topic)
+
+    return client
+
+
+def _create_username(hostname: str, client_id: str, product_info: str) -> str:
+    query_param_seq = []
+
+    # Apply query parameters (i.e. key1=value1&key2=value2...&keyN=valueN format)
+    if product_info.startswith(constant.DIGITAL_TWIN_PREFIX):  # Digital Twin Stuff
+        query_param_seq.append(("api-version", constant.IOTHUB_API_VERSION))
+        query_param_seq.append(("DeviceClientType", user_agent.get_iothub_user_agent()))
+        query_param_seq.append((constant.DIGITAL_TWIN_QUERY_HEADER, product_info))
+    else:
+        query_param_seq.append(("api-version", constant.IOTHUB_API_VERSION))
+        query_param_seq.append(
+            ("DeviceClientType", user_agent.get_iothub_user_agent() + product_info)
+        )
+
+    # NOTE: Client ID (including the device and/or module ids that are in it)
+    # is NOT url encoded as part of the username. Neither is the hostname.
+    # The sequence of key/value property pairs (query_param_seq) however, MUST have all
+    # keys and values URL encoded.
+    # See the repo wiki article for details:
+    # https://github.com/Azure/azure-iot-sdk-python/wiki/URL-Encoding-(MQTT)
+    username = "{hostname}/{client_id}/?{query_params}".format(
+        hostname=hostname,
+        client_id=client_id,
+        query_params=urllib.parse.urlencode(query_param_seq, quote_via=urllib.parse.quote),
+    )
+    return username
+
+
+def _create_c2d_message_generator(
+    device_id: str, mqtt_client: mqtt.MQTTClient
+) -> AsyncGenerator[Message, None]:
+    c2d_msg_topic = mqtt_topic.get_c2d_topic_for_subscribe(device_id)
+    mqtt_msg_generator = mqtt_client.get_incoming_message_generator(c2d_msg_topic)
+
+    async def c2d_message_generator(
+        incoming_mqtt_messages: AsyncGenerator[mqtt.MQTTMessage, None]
+    ) -> AsyncGenerator[Message, None]:
+        async for mqtt_message in incoming_mqtt_messages:
+            # TODO: decode differently depending on encoding type from topic
+            c2d_message = Message(mqtt_message.payload.decode("utf-8"))
```

Comment:
>c2d_message = Message(mqtt_message.payload.decode("utf-8")) [](http://example.com/codeflow?start=12&length=59)

If payload.decode raises an exception, what happens with this generator? Does it continue to function or do we need to somehow make a new one?

Reply:
If we add error handling logic in there, it'll continue to function